File: apps/t5_base_grammar_correction/iOS/T5GrammarCorrection_Project/T5GrammarCorrection/View/Tokenizer.swift
import Foundation

/// A placeholder Tokenizer. In a real app, implement SentencePiece or use a library.
/// T5 uses SentencePiece (spiece.model).
class Tokenizer {
    private var idToToken: [Int: String] = [:]
    private var tokenToId: [String: Int] = [:]
    
    init() {
        loadVocab()
    }
    
    private func loadVocab() {
        guard let url = Bundle.main.url(forResource: "t5_vocab", withExtension: "json") else {
            print("Tokenizer: t5_vocab.json not found in Bundle.")
            return
        }
        
        do {
            let data = try Data(contentsOf: url)
            let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: String] // Keys are usually strings in JSON
            
            // JSON keys are Strings "0", "1"... convert to Int
            if let json = json {
                for (key, value) in json {
                    if let id = Int(key) {
                        idToToken[id] = value
                        tokenToId[value] = id
                    }
                }
            }
            print("Tokenizer: Loaded \(idToToken.count) tokens.")
            
            // DEBUG: Verify critical tokens
            if let id = tokenToId["grammar"] { print("DEBUG: 'grammar' ID = \(id)") } else { print("DEBUG: 'grammar' NOT FOUND") }
            if let id = tokenToId["\u{2581}He"] { print("DEBUG: ' He' ID = \(id)") } else { print("DEBUG: ' He' NOT FOUND") }
            
        } catch {
            print("Tokenizer: Failed to load vocab - \(error)")
        }
    }
    
    // Optimized "Split & Merge" Encode for T5
    // 1. Split by space to get rough word boundaries.
    // 2. Prepend '▁' (\u{2581}) to non-start words.
    // 3. For each chunk, use greedy longest-prefix-match.
    func encode(_ text: String) -> [Int64] {
        var ids: [Int64] = []
        
        let words = text.split(separator: " ", omittingEmptySubsequences: false)
        for (i, wordSub) in words.enumerated() {
            var tokenString = String(wordSub)
            
            // T5 SentencePiece: words (except start) starting with space get the underscore prefix
            if i > 0 {
                tokenString = "\u{2581}" + tokenString
            }
            
            // Greedily consume this token string
            var remaining = tokenString
            // print("Tokenizer: Processing chunk '\(remaining)'") // Uncomment for deep debug
            while !remaining.isEmpty {
                var matchFound = false
                // Optimization: Check full string first (very common)
                if let id = tokenToId[remaining] {
                    ids.append(Int64(id))
                    // print("  -> Matched '\(remaining)' as \(id)")
                    remaining = ""
                    matchFound = true
                } else {
                    // Binary search or stride? Stride is fine for short fragments (words)
                    // limit max length check to remaining count
                    let maxLen = remaining.count
                    // We can cap at 20 or 30 if vocab max len is known, but maxLen is safe for words.
                    for len in stride(from: maxLen, through: 1, by: -1) {
                        let prefix = String(remaining.prefix(len))
                        if let id = tokenToId[prefix] {
                            ids.append(Int64(id))
                            remaining.removeFirst(len)
                            matchFound = true
                            break
                        }
                    }
                }
                
                if !matchFound {
                    // Unknown char fallback
                    // In real SP, backoff or byte fallback. Here, just UNK.
                    // Advance 1 char to avoid infinite loop
                    remaining.removeFirst()
                    ids.append(2) // <unk>
                }
            }
        }
        
        ids.append(1) // EOS
        return ids
    }

    func decode(_ tokens: [Int64]) -> String {
        var result = ""
        for (index, id) in tokens.enumerated() {
            if id == 1 { break } // EOS
            if let token = idToToken[Int(id)] {
                // T5 Token: " word" (U+2581 word) or "word"
                // We assume standard T5 spiece.model behavior
                var partial = token.replacingOccurrences(of: "\u{2581}", with: " ")
                
                // Fallback for literal escaped
                if partial.contains("\\u2581") {
                    partial = partial.replacingOccurrences(of: "\\\\u2581", with: " ")
                    partial = partial.replacingOccurrences(of: "\\u2581", with: " ")
                }

                result += partial
            }
        }
        
        var clean = result.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // Post-Processing for Punctuation (Common in T5 detokenization)
        // Fix "word , word" -> "word, word"
        clean = clean.replacingOccurrences(of: " ,", with: ",")
        clean = clean.replacingOccurrences(of: " .", with: ".")
        clean = clean.replacingOccurrences(of: " !", with: "!")
        clean = clean.replacingOccurrences(of: " ?", with: "?")
        clean = clean.replacingOccurrences(of: " '", with: "'")
        clean = clean.replacingOccurrences(of: " `", with: "'") // Fix backtick artifacts
        
        // Reduce double spaces if any
        while clean.contains("  ") {
            clean = clean.replacingOccurrences(of: "  ", with: " ")
        }
        
        return clean
    }
}

File: apps/t5_base_grammar_correction/Android/app/src/main/java/com/zeticai/t5grammar/Tokenizer.kt
package com.zeticai.t5grammar

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader

class Tokenizer(context: Context) {
    private val tokenToId = HashMap<String, Int>()
    private val idToToken = HashMap<Int, String>()

    init {
        loadVocab(context)
    }

    private fun loadVocab(context: Context) {
        try {
            val inputStream = context.assets.open("t5_vocab.json")
            val reader = BufferedReader(InputStreamReader(inputStream))
            val sb = StringBuilder()
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                sb.append(line)
            }
            reader.close()

            val json = JSONObject(sb.toString())
            val keys = json.keys()
            while (keys.hasNext()) {
                val key = keys.next() // ID as String
                val token = json.getString(key)
                val id = key.toInt()
                
                tokenToId[token] = id
                idToToken[id] = token
            }
            Log.d("Tokenizer", "Loaded ${tokenToId.size} tokens.")
        } catch (e: Exception) {
            Log.e("Tokenizer", "Failed to load vocab", e)
        }
    }

    // "Split & Match" Strategy matching Swift implementation
    fun encode(text: String): LongArray {
        val ids = ArrayList<Long>()
        
        // 1. Split by space
        val words = text.split(" ")
        
        for ((i, word) in words.withIndex()) {
            if (word.isEmpty()) continue
            
            var tokenString = word
            // T5: Prepend Lower One Eighth Block to non-start words
            if (i > 0) {
                tokenString = "\u2581" + tokenString
            }
            
            // Greedy Match
            var remaining = tokenString
            while (remaining.isNotEmpty()) {
                var matchFound = false
                
                // Optimization: Check full string first
                if (tokenToId.containsKey(remaining)) {
                    ids.add(tokenToId[remaining]!!.toLong())
                    remaining = ""
                    matchFound = true
                } else {
                    // Check prefixes
                    val maxLen = remaining.length
                    for (len in maxLen downTo 1) {
                        val prefix = remaining.substring(0, len)
                        if (tokenToId.containsKey(prefix)) {
                            ids.add(tokenToId[prefix]!!.toLong())
                            remaining = remaining.substring(len)
                            matchFound = true
                            break
                        }
                    }
                }
                
                if (!matchFound) {
                    // UNK
                    remaining = remaining.substring(1)
                    ids.add(2) // <unk>
                }
            }
        }
        
        ids.add(1) // EOS
        return ids.toLongArray()
    }

    fun decode(tokens: LongArray): String {
        val sb = StringBuilder()
        for (token in tokens) {
            if (token == 1L) break // EOS
            
            val tokenStr = idToToken[token.toInt()]
            if (tokenStr != null) {
                // T5 Tokenizer uses U+2581 (Lower One Eighth Block) for spaces
                // We replace it with a normal space.
                var cleanToken = tokenStr.replace("\u2581", " ")
                
                // Fallback: Check for literal escaped unicode if JSON parser didn't unescape
                if (cleanToken.contains("\\u2581")) {
                     cleanToken = cleanToken.replace("\\u2581", " ")
                }
                
                sb.append(cleanToken)
            }
        }
        
        var result = sb.toString().trim()
        
        // Post-Processing for Punctuation (Common in T5 detokenization)
        // Fix "word , word" -> "word, word"
        // Fix "word ." -> "word."
        result = result.replace(" ,", ",")
            .replace(" .", ".")
            .replace(" !", "!")
            .replace(" ?", "?")
            .replace(" '", "'") // Fix "don 't" -> "don't" tokens
            .replace(" `", "'") // Fix "don `t" -> "don't" artifacts
            
        // Reduce double spaces if any
        result = result.replace("  ", " ")
            
        return result
    }
}
